### PR TITLE
SISRP-10918 Rework sis-qa-test-auto dashboard

### PR DESCRIPTION
**Pull request by mDibyo**
https://jira.berkeley.edu/browse/SISRP-10918

### CODE REVIEW

**Comment by mDibyo** [comment]
qat dashboard
Screenshot of new dashboard interface

**Comment by mDibyo** [comment]
@christianvuerings Ready for review.

File: dashboard/.jshintrc [hunk 1]
```diff
@@ -4,15 +4,17 @@
...
   "freeze": true,
   "globals": {
-    "$": false
+    "angular": false
   },
   "immed": true,
   "indent": 2,
-  "latedef": true,
+  "jquery": true,
+  "latedef": "nofunc",
```
Comment (christianvuerings) [inline dashboard/.jshintrc:17]: @mDibyo could you explain a bit why we needed this?

File: dashboard/app/assets/javascripts/app.js [hunk 2]
```diff
@@ -0,0 +1,14 @@
+(function() {
+  'use strict';
+
+  angular.module('qat-dashboard', [
+    'ngResource',
+    'ui.router',
```
Comment (christianvuerings) [inline dashboard/app/assets/javascripts/app.js:6]: @mDibyo nice!

File: dashboard/app/assets/javascripts/dashboard.controller.js [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+
+  angular.module('qat-dashboard').controller('DashboardController', [
+    'testExecutor',
+    function(testExecutor) {
+      var vm = this;
+
+      vm.testExecsMap = testExecutor.all;
+      vm.requestTestExec = requestTestExec;
+      vm.getDate = getDate;
+      vm.getStatusCssClass = getStatusCssClass;
```
Comment (christianvuerings) [inline dashboard/app/assets/javascripts/dashboard.controller.js:12]: @mDibyo we probably want these at the end of the function. We should try to declare a function before it's used.

File: dashboard/app/assets/javascripts/dashboard.controller.js [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+      };
+
+      function requestTestExec() {
+        testExecutor.new();
+      }
+
+      function getDate(testExec) {
+        var uuid = testExec.uuid;
+
+        var year = parseInt(uuid.substring(0, 4));
```
Comment (christianvuerings) [inline dashboard/app/assets/javascripts/dashboard.controller.js]: @mDibyo make sure to always provide a radix parameter when using `parseInt`

File: dashboard/app/assets/javascripts/dashboard.controller.js [hunk 5]
```diff
@@ -0,0 +1,49 @@
+(function() {
+  'use strict';
+
+  angular.module('qat-dashboard').controller('DashboardController', [
+    'testExecutor',
+    function(testExecutor) {
+      var vm = this;
```
Comment (christianvuerings) [inline dashboard/app/assets/javascripts/dashboard.controller.js:7]: @mDibyo Instead of referencing `this`, I would recommend doing the following at the bottom:

```
return {
  testExecsMap: testExecutor.all,
  // ...
}
```

Let's do this in all of our files.

File: dashboard/app/assets/javascripts/dashboard.controller.js [hunk 6]
```diff
@@ -0,0 +1,49 @@
...
+      function getDate(testExec) {
+        var uuid = testExec.uuid;
+
+        var year = parseInt(uuid.substring(0, 4));
+        var month = parseInt(uuid.substring(4, 6));
+        var day = parseInt(uuid.substring(6, 8));
+        var hour = parseInt(uuid.substring(8, 10));
+        var minute = parseInt(uuid.substring(10, 12));
+        var second = parseInt(uuid.substring(12, 14));
+        var millisecond = parseInt(uuid.substring(14, 17));
```
Comment (christianvuerings) [inline dashboard/app/assets/javascripts/dashboard.controller.js]: @mDibyo not sure how you feel about this but would a `regex` work in this case?

File: dashboard/app/assets/javascripts/testExecutor.service.js [hunk 7]
```diff
@@ -0,0 +1,73 @@
...
+
+      function deleteTestExec(uuid) {
+        testExecsFactory.delete({id: uuid}, function(testExec) {
+          delete service.all[testExec];
+        });
+      }
+    }
+  ]);
+
+  angular.module('qat-dashboard').factory('testExecsFactory', [
```
Comment (christianvuerings) [inline dashboard/app/assets/javascripts/testExecutor.service.js]: @mDibyo could we put this in a separate file? 1 file per factory / service.

File: dashboard/app/assets/javascripts/toArray.filter.js [hunk 8]
```diff
@@ -0,0 +1,31 @@
+(function() {
+  'use strict';
+
+  angular.module('qat-dashboard').filter('toArray', function() {
```
Comment (christianvuerings) [inline dashboard/app/assets/javascripts/toArray.filter.js:4]: @mDibyo instead of copy / pasting this in our own repo, could we use the npm package instead?
https://www.npmjs.com/package/angular-toarrayfilter

File: dashboard/app/assets/stylesheets/dashboard.scss [hunk 9]
```diff
@@ -0,0 +1,57 @@
...
+
+$dashboard-item-status-width: 120px;
+
+.dashboard-heading {
+  padding-left: 25px;
+}
+
+.dashboard-item {
+  position: relative;
+  width: 900px;
```
Comment (christianvuerings) [inline dashboard/app/assets/stylesheets/dashboard.scss:11]: @mDibyo do you know whether this will ultimately be responsive? Would be great if we could look at this from our mobile phones.

File: dashboard/app/assets/stylesheets/dashboard.scss [hunk 10]
```diff
@@ -0,0 +1,57 @@
...
+  top: 0;
+  width: $dashboard-item-status-width;
+
+  @each $status, $color in
+  (queued, #bdbdbd),
+  (dryrun, #26c6da),
+  (executing, #7e57c2),
+  (done, #66bb6a),
+  (done-with-failed, #ffca28),
+  (errored, #ef5350) {
```
Comment (christianvuerings) [inline dashboard/app/assets/stylesheets/dashboard.scss]: @mDibyo beautiful, thanks!

File: dashboard/app/assets/stylesheets/util.scss [hunk 11]
```diff
@@ -0,0 +1,4 @@
+@mixin font-main($color, $weight) {
```
Comment (christianvuerings) [inline dashboard/app/assets/stylesheets/util.scss]: @mDibyo I do like mixins, but let's only use them when they do tricky things like calculations.
Could we remove this one?

File: dashboard/package.json [hunk 12]
```diff
@@ -20,12 +20,18 @@
   "bugs": {
     "url": "https://github.com/ucberkeley/sis-qa-test-auto/issues"
   },
-  "devDependencies": {
+  "dependencies": {
+    "browserify": "~13.0.0",
```
Comment (christianvuerings) [inline dashboard/package.json:24]: @mDibyo is there a reason we put them in different sections - `devDependencies` and `dependencies`? Could we combine them instead?

File: dashboard/public/html/dashboard.html [hunk 13]
```diff
@@ -0,0 +1,60 @@
+<md-content>
+  <md-toolbar class="md-medium-tall md-whiteframe-18dp" style="position: fixed;" layout-align="center">
```
Comment (christianvuerings) [inline dashboard/public/html/dashboard.html]: @mDibyo could we put this inline style in a `class`?

File: dashboard/public/html/dashboard.html [hunk 14]
```diff
@@ -0,0 +1,60 @@
...
+    </div>
+    <md-card
+        class="dashboard-item"
+        ng-repeat="testExec in vm.testExecsMap | toArray | orderBy:'uuid':true track by testExec.uuid"
+        layout="row" layout-align="center center" layout-wrap
+        md-ink-ripple>
+      <md-card-content flex>
+        <div layout="row">
+            <span class="dashboard-item-uuid-time">
+              {{vm.getDate(testExec) | date:'medium'}}
```
Comment (christianvuerings) [inline dashboard/public/html/dashboard.html]: @mDibyo could we use the `ng-bind` approach instead of using curly brackets?

File: dashboard/public/html/dashboard.html [hunk 15]
```diff
@@ -0,0 +1,60 @@
...
+            </span>
+            <span class="dashboard-item-counters" flex>
+              <span ng-hide="testExec.status==='QUEUED'||testExec.status==='DRYRUN'">
+                <div>
+                  TOTAL:
+                  <span class="dashboard-item-counters-count">{{testExec.counters.total}}</span>
+                </div>
+                &nbsp;&nbsp;&nbsp;
```
Comment (christianvuerings) [inline dashboard/public/html/dashboard.html]: @mDibyo this seems like a hack, is there no better way to achieve this (e.g. margin / padding in CSS)

File: dashboard/public/index.html [hunk 16]
```diff
@@ -0,0 +1,27 @@
...
+<html>
+
+<head>
+  <title>QAT Dashboard</title>
+
+  <link rel='stylesheet' type='text/css' href='https://fonts.googleapis.com/css?family=Roboto:400,700'>
+  <link rel="stylesheet" href="http://ajax.googleapis.com/ajax/libs/angular_material/1.0.0/angular-material.min.css">
+  <link rel="stylesheet" href="css/app.css">
+
+  <script src="//ajax.googleapis.com/ajax/libs/angularjs/1.4.8/angular.min.js"></script>
```
Comment (christianvuerings) [inline dashboard/public/index.html:12]: @mDibyo if there is a way we can load these as `npm` modules, that would be great, that way we have all our external front-end dependencies in 1 file.

**Comment by christianvuerings** [comment]
@mDibyo left some small comments inline but overall I have to say this looks awesome! Thank you.

File: dashboard/app/assets/javascripts/dashboard.controller.js (hunk 5 above)
Comment (mDibyo) [inline dashboard/app/assets/javascripts/dashboard.controller.js:7]: @christianvuerings This would results in failing of instanceof tests: http://stackoverflow.com/questions/2938940/return-statement-in-js-constructors

Would that be ok?

File: dashboard/app/assets/javascripts/dashboard.controller.js (hunk 4 above)
Comment (mDibyo) [inline dashboard/app/assets/javascripts/dashboard.controller.js]: Done.

File: dashboard/app/assets/javascripts/dashboard.controller.js (hunk 6 above)
Comment (mDibyo) [inline dashboard/app/assets/javascripts/dashboard.controller.js]: That's a great idea! Done.

File: dashboard/app/assets/javascripts/testExecutor.service.js (hunk 7 above)
Comment (mDibyo) [inline dashboard/app/assets/javascripts/testExecutor.service.js]: @christianvuerings I placed testExecutorFactory in this file because it was supposed to be used only in the testExecutor service. I have deleted the factory and moved its functionality into the testExecutor service.

File: dashboard/app/assets/javascripts/dashboard.controller.js (hunk 4 above)
Comment (mDibyo) [inline dashboard/app/assets/javascripts/dashboard.controller.js]: Done.

File: dashboard/app/assets/javascripts/dashboard.controller.js (hunk 5 above)
Comment (christianvuerings) [inline dashboard/app/assets/javascripts/dashboard.controller.js:7]: @mDibyo I think that should be fine. With Angular we usually don't use that pattern.

Comment (mDibyo) [inline dashboard/app/assets/javascripts/dashboard.controller.js:7]: @christianvuerings To add to my note above, I was following this styleguide: https://github.com/johnpapa/angular-styleguide. Specifically, the following sections:
- https://github.com/johnpapa/angular-styleguide#controlleras-with-vm
- https://github.com/johnpapa/angular-styleguide#bindable-members-up-top

`var vm = this;` allows us to access `this` inside a function. Binding functions at the top (before defining helps with readability. )

I have since found, that I don't have to access `this` anywhere within functions if I am good with my variable declarations. And I can preserve the instanceof behavior by using the following pattern:

```
angular.extend(this, {
  testExecsMap: testExecutor.all,
  // ...
})
```

I will move the code to this pattern.

File: dashboard/.jshintrc (hunk 1 above)
Comment (mDibyo) [inline dashboard/.jshintrc:17]: @christianvuerings Please check my explanation below. The relevant portion is discussed here: https://github.com/johnpapa/angular-styleguide#bindable-members-up-top

File: dashboard/app/assets/javascripts/dashboard.controller.js (hunk 3 above)
Comment (mDibyo) [inline dashboard/app/assets/javascripts/dashboard.controller.js:12]: @christianvuerings Please check my explanation below. The relevant portion is discussed here: https://github.com/johnpapa/angular-styleguide#bindable-members-up-top

File: dashboard/public/index.html (hunk 16 above)
Comment (mDibyo) [inline dashboard/public/index.html:12]: @christianvuerings I have been trying not to load in standard libraries from npm, instead relying on the browser/CDN to get cached versions. This would speed up first load for the dashboard, but will have no impact on subsequent loads (since everything will be cached. ) Do you think I should shift to loading from npm modules?

File: dashboard/package.json (hunk 12 above)
Comment (mDibyo) [inline dashboard/package.json:24]: @christianvuerings I have placed the test dependencies into 'devDependencies' since we presumably don't need them in production environments. In production, only the stuff in 'dependencies' need to be installed. What do you think?

File: dashboard/public/index.html (hunk 16 above)
Comment (christianvuerings) [inline dashboard/public/index.html:12]: @mDibyo You're correct that it might be faster on first load, but having our dependencies in 1 file outweighs that. So let's use `npm` modules for our external dependencies for now.

File: dashboard/app/assets/javascripts/dashboard.controller.js (hunk 5 above)
Comment (christianvuerings) [inline dashboard/app/assets/javascripts/dashboard.controller.js:7]: @mDibyo after reading those through, I'm ok either way. Thanks for clarifying.

File: dashboard/package.json (hunk 12 above)
Comment (christianvuerings) [inline dashboard/package.json:24]: @mDibyo Sounds good, thanks for clarifying. We just need to make sure we keep that line of thought moving forward.

File: dashboard/app/assets/stylesheets/util.scss (hunk 11 above)
Comment (mDibyo) [inline dashboard/app/assets/stylesheets/util.scss]: @christianvuerings That makes sense. Done.
